src/AqueductV1Pair.sol

```diff
@@ -326,224 +442,244 @@ contract AqueductV1Pair is IAqueductV1Pair, AqueductV1ERC20, SuperAppBase {
+            _updateAccumulators(reserve0, reserve1, totalFlow0, totalFlow1, time);
 
-            _totalSwappedFunds0 = totalSwappedFunds0; // gas savings
-            _totalSwappedFunds1 = totalSwappedFunds1;
-            balance0 = IERC20(_token0).balanceOf(address(this)) - _totalSwappedFunds0;
-            balance1 = IERC20(_token1).balanceOf(address(this)) - _totalSwappedFunds1;
+            totalSwappedFunds0 = _totalSwappedFunds0; // gas savings
+            totalSwappedFunds1 = _totalSwappedFunds1;
+            balance0 = IERC20(_token0).balanceOf(address(this)) - totalSwappedFunds0;
+            balance1 = IERC20(_token1).balanceOf(address(this)) - totalSwappedFunds1;
 
-            (_reserve0, _reserve1) = _getReservesAtTime(time, totalFlow0, totalFlow1);
-            feeOn = _mintFee(_reserve0, _reserve1);
+            feeOn = _mintFee(reserve0, reserve1);
         }
 
-        uint256 _totalSupply = totalSupply; // gas savings, must be defined here since totalSupply can update in _mintFee
+        // gas savings, must be defined here since totalSupply can update in _mintFee
+        uint256 _totalSupply = totalSupply;
         amount0 = (liquidity * balance0) / _totalSupply; // using balances ensures pro-rata distribution
         amount1 = (liquidity * balance1) / _totalSupply; // using balances ensures pro-rata distribution
         if (amount0 <= 0 || amount1 <= 0) revert PAIR_INSUFFICIENT_LIQUIDITY_BURNED();
         _burn(address(this), liquidity);
         _safeTransfer(_token0, to, amount0);
         _safeTransfer(_token1, to, amount1);
-        balance0 = IERC20(_token0).balanceOf(address(this)) - _totalSwappedFunds0;
-        balance1 = IERC20(_token1).balanceOf(address(this)) - _totalSwappedFunds1;
+        balance0 = IERC20(_token0).balanceOf(address(this)) - totalSwappedFunds0;
+        balance1 = IERC20(_token1).balanceOf(address(this)) - totalSwappedFunds1;
 
         _updateReserves(balance0, balance1, time);
-        if (feeOn) kLast = uint256(reserve0) * reserve1; // reserve0 and reserve1 are up-to-date
+        if (feeOn) kLast = uint256(_reserve0) * _reserve1; // _reserve0 and _reserve1 are up-to-date
         emit Burn(msg.sender, amount0, amount1, to);
     }
 
-    // this low-level function should be called from a contract which performs important safety checks
-    function swap(uint256 amount0Out, uint256 amount1Out, address to, bytes calldata data) external override lock {
+    /**
+     * @notice Swap tokens for the desired amounts.
+     * @dev This low-level function should be called from a contract which performs important safety checks
+     * @param amount0Out The amount of token0 to be swapped.
+     * @param amount1Out The amount of token1 to be swapped.
+     * @param to The recipient address of the swapped tokens.
+     */
+    function swap(uint256 amount0Out, uint256 amount1Out, address to) external override lock {
+        if (msg.sender != factory) revert PAIR_FORBIDDEN(); // TODO: is this ok?
```

Comment:
Calling from the factory feels a little wrong as the factory's responsibility should be to create pools.

Was there a reason for putting the logic in the factory?

Would a better separation of concerns be: 
- Have the auction in a dedicated auction contract
- Have the swap function called by `AqueductV1Router.sol`
- Then also call the auction contract through the router. The router has access to the factor so you should be able to access all the same stuff (unless you're using some private vars/functions in the factory).

Thoughts?

Reply:
Another thought, can you remind me why we need the line in the first place? Or why we need to check the swap function is called from a separate contract:

> if (msg.sender != factory) revert PAIR_FORBIDDEN();

Just curious as uniswap v2 doesn't do a similar check

Reply:
the swap function can only be called by the auction contract, in this case is the factory. Otherwise of course traders would be able to bypass the auction and swap directly at the lowest fee